config/config.go

```diff
@@ -169,6 +174,18 @@ func (tree *Tree) Lookup(abspath project.Path) (*Tree, bool) {
 	return cfg, true
 }
 
+// Dirs returns all the configuration directories that have configurations.
```

Comment:
There is some stuttering here, thinking if we can't just assume that subdirs/dirs inside the tree always have configurations, because our tree is a config tree already right ? unless we do have some dirs sometimes that have no config at all, but if it represents only configuration it is already an invariant that anything inside has configs and is config related etc (it is not a fs tree). But I'm not sure, just thinking out loud (also not fundamental or anything).

If what Im saying is not completely senseless, maybe this could be just:

```suggestion
// Dirs returns a list with this configuration and all its children.
```

Now not so sure about calling this Dirs...you are actually getting the whole tree as a list no ? Maybe Collect ? Then it would be: "Collect returns the entire tree collected as a depth first list".

Sorry if this is all stupid, feel free to disregard.

Reply:
The docs in stupid, thanks for catching this.

> There is some stuttering here, thinking if we can't just assume that subdirs/dirs inside the tree always have configurations, because our tree is a config tree already right ? 

Currently the `config.Tree` maintains the same filesystem structure of the project, then directories with no configuration are kept in the tree, then `node.Parent` points to the parent directory (if any).

But if we change to ignore the directories with no configuration in the tree, then maybe with the gaps it will not be enough to express all use cases. 
Eg.:
```
s:stacks
s:stacks/orgs/mineiros
s:stacks/orgs/google
s:stacks/orgs/meta
```
The `stacks/orgs` directory has no configuration, then in this case, it will not be possible to call `generate.Do(root.Lookup("/stacks/orgs"))` to execute generate from that directory.

That's why (I think) we should keep the configuration tree resembling the filesystem tree, because sometimes we want to execute something using the *view* of some directory (eg: the working dir filter).

Each Tree is actually the directory, and it can hold a config or not... the `dir` is ommited in the Tree name but it's there. I just thought that calling it `config.Dir` would be funny.

But I agree calling this `Dirs()` but only returning dirs with config is not correct.
What about `cfg.AsList()`? and then it returns all dir configs and the client filters out by `!cfg.IsEmpty()`?

```suggestion
// AsList returns a list with this configuration and all its children.
```

Reply:
I think I was thinking on something like the skipdir feature, today because of skipdir we already don't load the whole fs tree, so it is not a perfect representation, but makes sense that a lot of nodes will be empty.

AsList should do the trick, I prefer Collect because AsX always reminds me of casting, it looks like you are just to cast something to a different type (like errors.As()), but it is something only on my head/subjective and totally ignorable.